test/server.ts

```diff
@@ -237,6 +244,135 @@ test('store server started in the background should use store location wanted by
   t.notOk(await pathExists(serverJsonPath), 'server.json removed')
 })
 
+interface ChildProcess extends EventEmitter {
+  pid: number;
+  stderr: Readable;
+  stdout: Readable;
+}
+
+type ServerProcessList = Array<[ChildProcess, Deferred<void>]>
+
+async function testParallelServerStart (options: {
+  onTimeout?: (serverProcessList: ServerProcessList) => Promise<void>,
+  timeoutMillis?: number,
+  onProcessStarted?: (serverProcess: ChildProcess, serverProcessDeferred: Deferred<void>) => void,
+  onProcessClosed?: (serverProcess: ChildProcess) => void,
+  }) {
+  const pnpmBin = 'pnpm'
+
+  // List of pairs of ChildProcess, Deferred
+  const serverProcessList: ServerProcessList = []
+
+  let timeoutObj: number|NodeJS.Timer|null = setTimeout(() => {
+    timeoutObj = null
+    completedPromise = completedPromise.then(() => {
+      if (options.onTimeout) {
+        options.onTimeout(serverProcessList)
+      }
+    })
+    for (const [serverProcess, serverProcessDeferred] of serverProcessList) {
+      kill(serverProcess.pid, 'SIGINT')
+      serverProcessDeferred.resolve()
+    }
+  }, !options.timeoutMillis && options.timeoutMillis !== 0 ? 10000 : options.timeoutMillis)
+  let completedPromise: Promise<void> = Promise.resolve()
+  for (let i = 0; i < 10; i++) {
+    const serverProcess = crossSpawn.spawn(
+      pnpmBin,
+      ['server', 'start',
+        '--store', '../store2',
+        // We don't have to specify --port or --path explicitly because server.json exclusive access also fixes
+        // the possibility of port collision.
+      ])
+    const serverProcessDeferred = createDeferred<void>()
+    serverProcessList.push([serverProcess, serverProcessDeferred])
+    if (options.onProcessStarted) {
+      (options.onProcessStarted)(serverProcess, serverProcessDeferred)
+    }
+    serverProcess.on('close', (code: number) => {
+      if (options.onProcessClosed) {
+        (options.onProcessClosed)(serverProcess)
+      }
+      for (let j = 0; j < serverProcessList.length; j++) {
+        if (serverProcessList[j][0] === serverProcess) {
+          serverProcessList.splice(j, 1)
+          break
+        }
+      }
+      if (code !== 0) {
+        serverProcessDeferred.reject(new Error(`A server process unexpectedly exited with code ${code}`))
+      } else {
+        serverProcessDeferred.resolve()
+      }
+    });
+    completedPromise = completedPromise.then(() => serverProcessDeferred.promise)
+  }
+  await completedPromise
+  if (timeoutObj !== null) {
+    clearTimeout(timeoutObj)
+  }
+}
+
+test('parallel server starts against the same store should result in a winner within 10 seconds', async (t: tape.Test) => {
+  await testParallelServerStart({
+    onTimeout: async (serverProcessList: ServerProcessList) => {
+      if (serverProcessList.length > 1) {
+        throw new Error('There are still multiple server processes running after 10 seconds')
```

Comment:
You can use `t.fail('There are still multiple server processes running after 10 seconds')`

but I think the right assert is
`t.equal(serverProcessList.length, 1, 'one server is running after 10 seconds')`